File: packages/core/src/result.ts
/**
 * Result type — explicit success/failure without throwing.
 *
 * Inspired by Rust's Result and used widely at Stripe/Google-style codebases
 * to make error paths first-class. Prefer `Result` for expected/recoverable
 * outcomes; reserve `throw` for truly exceptional, unrecoverable cases.
 */

export type Ok<T> = { readonly ok: true; readonly value: T };
export type Err<E> = { readonly ok: false; readonly error: E };
export type Result<T, E = Error> = Ok<T> | Err<E>;

export function ok<T>(value: T): Ok<T> {
  return { ok: true, value };
}

export function err<E>(error: E): Err<E> {
  return { ok: false, error };
}

export function isOk<T, E>(r: Result<T, E>): r is Ok<T> {
  return r.ok;
}

export function isErr<T, E>(r: Result<T, E>): r is Err<E> {
  return !r.ok;
}

/** Return the value or throw the error. Use at boundaries only. */
export function unwrap<T, E>(r: Result<T, E>): T {
  if (r.ok) return r.value;
  throw r.error instanceof Error ? r.error : new Error(String(r.error));
}

/** Return the value or a fallback. */
export function unwrapOr<T, E>(r: Result<T, E>, fallback: T): T {
  return r.ok ? r.value : fallback;
}

/** Map the success value, leaving errors untouched. */
export function map<T, U, E>(r: Result<T, E>, fn: (value: T) => U): Result<U, E> {
  return r.ok ? ok(fn(r.value)) : r;
}

/** Wrap a throwing function into a Result. */
export function fromThrowable<T>(fn: () => T): Result<T, Error> {
  try {
    return ok(fn());
  } catch (e) {
    return err(e instanceof Error ? e : new Error(String(e)));
  }
}

/** Wrap a promise into a Result, never rejecting. */
export async function fromPromise<T>(promise: Promise<T>): Promise<Result<T, Error>> {
  try {
    return ok(await promise);
  } catch (e) {
    return err(e instanceof Error ? e : new Error(String(e)));
  }
}

File: python/getexp-core/getexp_core/result.py
"""Result type — explicit success/failure without raising.

Mirrors ``@getexp/core``'s Result. Prefer it for expected/recoverable
outcomes; reserve exceptions for truly exceptional cases.
"""

from __future__ import annotations

from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from typing import Generic, TypeVar

T = TypeVar("T")
U = TypeVar("U")
E = TypeVar("E")


@dataclass(frozen=True, slots=True)
class Ok(Generic[T]):
    value: T

    @property
    def ok(self) -> bool:
        return True


@dataclass(frozen=True, slots=True)
class Err(Generic[E]):
    error: E

    @property
    def ok(self) -> bool:
        return False


Result = Ok[T] | Err[E]


def ok(value: T) -> Ok[T]:
    return Ok(value)


def err(error: E) -> Err[E]:
    return Err(error)


def is_ok(r: Result[T, E]) -> bool:
    return isinstance(r, Ok)


def is_err(r: Result[T, E]) -> bool:
    return isinstance(r, Err)


def unwrap(r: Result[T, E]) -> T:
    """Return the value or raise the error. Use at boundaries only."""
    if isinstance(r, Ok):
        return r.value
    raise r.error if isinstance(r.error, BaseException) else RuntimeError(str(r.error))


def unwrap_or(r: Result[T, E], fallback: T) -> T:
    return r.value if isinstance(r, Ok) else fallback


def map_ok(r: Result[T, E], fn: Callable[[T], U]) -> Result[U, E]:
    return ok(fn(r.value)) if isinstance(r, Ok) else r


def from_callable(fn: Callable[[], T]) -> Result[T, Exception]:
    try:
        return ok(fn())
    except Exception as exc:
        return err(exc)


async def from_awaitable(awaitable: Awaitable[T]) -> Result[T, Exception]:
    try:
        return ok(await awaitable)
    except Exception as exc:
        return err(exc)
